lgc/patch/NggPrimShader.cpp

```diff
@@ -3356,84 +3163,169 @@ Function *NggPrimShader::mutateGsToVariant(Module *module) {
 
   m_builder->restoreIP(savedInsertPos);
 
-  return gsEntryVariant;
+  return gsEntryPoint;
 }
 
 // =====================================================================================================================
 // Runs copy shader.
 //
 // @param module : LLVM module
-// @param insertAtEnd : Where to insert instructions
-void NggPrimShader::runCopyShader(Module *module, BasicBlock *insertAtEnd) {
+void NggPrimShader::runCopyShader(Module *module) {
   assert(m_hasGs); // GS must be present
 
-  auto copyShaderEntryPoint = module->getFunction(lgcName::NggCopyShaderEntryPoint);
+  //
+  // The processing is something like this:
+  //
+  //   uncompactVertexId = Thread ID in subgroup
+  //   if (vertCompacted)
+  //     uncompactVertexId = Read uncompacted vertex ID from LDS
+  //   Calculate vertex offset and run copy shader
+  //
+  auto expVertBlock = m_builder->GetInsertBlock();
+
+  auto reverseMapVertThreadIdBlock = createBlock(expVertBlock->getParent(), ".reverseMapVertThreadId");
+  reverseMapVertThreadIdBlock->moveAfter(expVertBlock);
 
-  // Mutate copy shader entry-point, handle GS output import
+  auto expVertContinueBlock = createBlock(expVertBlock->getParent(), ".expVertContinue");
+  expVertContinueBlock->moveAfter(reverseMapVertThreadIdBlock);
+
+  auto vertCompacted = m_builder->CreateICmpULT(m_nggFactor.vertCountInSubgroup, m_nggFactor.maxVertOutInSubgroup);
+  m_builder->CreateCondBr(vertCompacted, reverseMapVertThreadIdBlock, expVertContinueBlock);
```

Comment:
This branch seems to be redundant with the equivalent branch in `constructPrimShader`.

Reply:
This is to avoid query vertex ID map if you don't need vertex compaction. The condition is based on this `actualNumEmittedVertices < expectedNumEmittedVertices`. With this condition, reconstruction of primitive connectivity data is unnecessary.

Reply:
Got it, thanks!